jupyterhub/tests/selenium/test_browser.py

```diff
@@ -0,0 +1,417 @@
+
+    if location:
+        location = ujoin(app.base_url, location)
+        print("if location is:" + location)
+    elif running:
+        # location not specified,
+        location = user.url
+        if params:
+            location = url_concat(location, params)
+    else:
+        # use default url
+        location = ujoin(app.base_url, 'hub/spawn')
+        if params:
+            location = url_concat(location, params)
+
+    url = 'login'
+    if params:
+        url = url_concat(url, params)
+    if next_url:
+        if next_url.startswith('/') and not (
+            next_url.startswith("//") or urlparse(next_url).netloc
+        ):
+            next_url = ujoin(app.base_url, next_url, '')
+        url = url_concat(url, dict(next=next_url))
+
+    if running and not user.active:
+        # ensure running
+        await user.spawn()
+    elif user.active and not running:
+        # ensure not running
+        await user.stop()
+    time.sleep(5)
+    # open_url(app,browser)
+    await in_thread(browser.get, (ujoin(public_url(app), url)))
+    await login(browser, user='test_user', pass_w='test_user')
+
+    # await get_page(url, app, allow_redirects=False)
+    time.sleep(50)
+
+    if user != 'admin' and next_url == '/hub/admin':  # user!=user.admin
+        assert browser.current_url.endswith('hub/admin/')
+        elm = browser.find_element(
+            LoginPageLocators.ERROR_403
+        )  # .get_attribute('action')
+        print(elm)
+        assert elm == LoginPageLocators.ERROR_MESSAGES_403
+        if params:
+            assert browser.current_url.endswith('hub/admin/')
+            elm = browser.find_element(
+                *LoginPageLocators.ERROR_403
+            )  # .get_attribute('action')
+            print(elm)
+            assert elm == LoginPageLocators.ERROR_MESSAGES_403
+
+    elif next_url == '/has?repo=https':
+        # print(app.users['test_user'].keys())
+        print("else if")
+    else:
+        print("else")
+
+    """
+    <div class="error">
+
+
+</div>
+    0. 1., 9-14, 17-18 /space word/user/test_user/
+    1. /space word/user/test_user/
+    2. - 15,16 - space word/hub/admin  <h1>
+    403 : Forbidden
+  </h1>: <h1>
+    <p>
+    Action is not authorized with current scopes; requires any of [admin-ui]
+  </p>
+
+    3. space word/hub/user/other   <h1>
+    404 : Not Found
+  </h1> <p>Jupyter has lots of moons, but this is not one...</p>
+    4. - 8.  the same as 3
+    """
+
+    """
+    print("url ---" +browser.current_url)
+    
+    #print("user ()" +f"{user}")
+    print("user (type)" +str(user))
+    assert location in browser.current_url      """
+
+
+@pytest.mark.parametrize(
+    "user, pass_w",
+    [
+        (" ", ""),
+        ("user", ""),
+        (" ", "password"),
+        ("user", "password"),
+    ],
+)
+async def test_invalid_credantials(app, browser, user, pass_w):
+
+    await open_url(app, browser)
+    await login(browser, user, pass_w)
+    time.sleep(5)
```

Comment:
we shouldn't have any time.sleep in the tests because they don't allow the server to respond. If anything, they should be `await asyncio.sleep`. But ideally, we should always wait for a specific condition. Sleeps always have a chance to be too short _sometimes_.